File: go/openai_adapter.go
package hydra

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

const defaultOpenAIBaseURL = "https://api.openai.com/v1"

type OpenAIAdapter struct {
	apiKey, model, baseURL string
}

func NewOpenAIAdapter(apiKey, model, baseURL string) *OpenAIAdapter {
	if baseURL == "" {
		baseURL = defaultOpenAIBaseURL
	}
	baseURL = strings.TrimRight(baseURL, "/")
	return &OpenAIAdapter{
		apiKey:  apiKey,
		model:   model,
		baseURL: baseURL,
	}
}

type openAIChatRequest struct {
	Model    string          `json:"model"`
	Messages []openAIMessage `json:"messages"`
}

type openAIMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type openAIChatResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Usage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
	} `json:"usage"`
}

func (a *OpenAIAdapter) Decompose(topic string, breadth int) (*DecomposerResponse, error) {
	tmpl, err := loadDecomposePrompt()
	if err != nil {
		return nil, err
	}
	prompt := strings.Replace(tmpl, "{{BREADTH}}", strconv.Itoa(breadth), 1)
	prompt = strings.Replace(prompt, "{{TOPIC}}", topic, 1)

	reqBody := openAIChatRequest{
		Model: a.model,
		Messages: []openAIMessage{
			{Role: "user", Content: prompt},
		},
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	endpoint := a.baseURL + "/chat/completions"
	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+a.apiKey)

	start := time.Now()
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	latency := int(time.Since(start).Milliseconds())

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API error (%d): %s", resp.StatusCode, string(body))
	}

	var chatResp openAIChatResponse
	if err := json.Unmarshal(body, &chatResp); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if len(chatResp.Choices) == 0 {
		return nil, fmt.Errorf("empty response: no choices returned")
	}

	text := chatResp.Choices[0].Message.Content

	var data struct {
		Subtopics []string `json:"subtopics"`
	}
	if err := ExtractJSON(text, &data); err != nil {
		return nil, fmt.Errorf("failed to extract JSON from response: %w", err)
	}

	if len(data.Subtopics) > breadth {
		data.Subtopics = data.Subtopics[:breadth]
	}

	return &DecomposerResponse{
		Subtopics: data.Subtopics,
		Metadata: NodeMetadata{
			Tokens:    chatResp.Usage.PromptTokens + chatResp.Usage.CompletionTokens,
			Model:     a.model,
			LatencyMS: latency,
		},
	}, nil
}

func (a *OpenAIAdapter) GetModelName() string {
	return a.model
}

File: src/openai_adapter.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { extractJson } from "./json_repair";
import type { Adapter, DecomposerResponse } from "./types";

interface OpenAIConfig {
  apiKey: string;
  model: string;
  baseUrl?: string;
}

interface ChatCompletionResponse {
  choices: Array<{
    message: {
      content: string | null;
    };
  }>;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
  };
}

export class OpenAIAdapter implements Adapter {
  private apiKey: string;
  private model: string;
  private baseUrl: string;

  constructor(config: OpenAIConfig) {
    this.apiKey = config.apiKey;
    this.model = config.model;
    this.baseUrl = (config.baseUrl ?? "https://api.openai.com/v1").replace(/\/+$/, "");
  }

  async decompose(topic: string, breadth: number): Promise<DecomposerResponse> {
    const tmpl = readFileSync(resolve(__dirname, "..", "protocol", "decompose.prompt"), "utf-8");
    const prompt = tmpl.replace("{{BREADTH}}", String(breadth)).replace("{{TOPIC}}", topic);

    const start = Date.now();

    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: this.model,
        max_tokens: 1024,
        messages: [{ role: "user", content: prompt }],
      }),
    });

    const latency = Date.now() - start;

    if (!response.ok) {
      const body = await response.text();
      throw new Error(`OpenAI API error ${response.status}: ${body.slice(0, 300)}`);
    }

    const result: ChatCompletionResponse = await response.json();

    const text = result.choices[0]?.message?.content ?? "";
    if (!text) {
      throw new Error("OpenAI API returned empty content");
    }

    const data = extractJson<{ subtopics: string[] }>(text);

    const tokens = (result.usage?.prompt_tokens ?? 0) + (result.usage?.completion_tokens ?? 0);

    return {
      subtopics: data.subtopics.slice(0, breadth),
      metadata: {
        tokens,
        model: this.model,
        latency_ms: latency,
      },
    };
  }

  getModelName(): string {
    return this.model;
  }
}
